sb_vision/distance_finding.py

```diff
@@ -0,0 +1,121 @@
+location finding function.
+"""
+
+import functools
+import re
+from pathlib import Path
+from typing import Any, Dict, List, NamedTuple, Tuple
+
+import cv2
+import numpy as np
+from lxml import etree
+
+Coordinate = NamedTuple('PixelCoordinate', [('x', float), ('y', float)])
+
+
+def _get_values_from_xml_element(element: etree.Element) -> List[str]:
+    """Parse an xml tag with space-separated variables."""
+    text = []  # type: List[str]
+    for e in element.itertext():
+        if e.strip():
+            text += re.split('\s+', re.sub('\n', '', e.strip()))
+    return text
+
+
+def get_calibration(file_name: Path) -> Dict[str, Any]:
+    """
+    Parse a calibration xml generated by the camera calibration tool.
+
+    (see https://docs.opencv.org/3.4.0/d7/d21/tutorial_interactive_calibration.html)
+    :param file_name: name of xml file to parse
+    :return: calibrations from the file in a dictionary
+    """
+    calibrations = {}
+    with file_name.open() as file:
+        tree = etree.parse(file)
+        root = tree.getroot()
+        for element in root:
+            if 'type_id' in element.attrib and element.attrib[
+               'type_id'] == 'opencv-matrix':
+                rows, cols = int(element.find('rows').text), int(
+                    element.find('cols').text)
+                data_type = element.find('dt').text
+                values = _get_values_from_xml_element(element.find('data'))
+                if data_type == 'd':  # doubles
+                    data = np.reshape([float(v) for v in values],
+                                      (rows, cols)).tolist()
+                else:
+                    raise ValueError('Invalid data type in xml')
+            # Integer tag names
+            elif element.tag in ['framesCount', 'cameraResolution']:
+                values = _get_values_from_xml_element(element)
+                data = [int(v) for v in values]
+            elif element.tag in ['avg_reprojection_error']:
+                values = _get_values_from_xml_element(element)
+                data = [float(v) for v in values]
+            else:
+                data = _get_values_from_xml_element(element)
+            calibrations[element.tag] = data
+    return calibrations
+
+
+@functools.lru_cache()
+def load_camera_calibrations(file_name: Path) -> Tuple[List[List[float]],
+                                                       List[List[float]]]:
+    """
+    Load camera calibrations from a file.
+
+    :param file_name: file to load
+    :return: camera calibrations
+    """
+    calibrations = get_calibration(file_name)
+    camera_matrix = calibrations['cameraMatrix']
+    distance_coefficents = calibrations['dist_coeffs']
+    return camera_matrix, distance_coefficents
+
+
+def calculate_transforms(
+    marker_size: Tuple[float, float],
+    pixel_coords: List[Coordinate],
+    camera_matrix: List[List[float]],
+    distance_coefficients: List[List[float]],
+):
```

Comment:
Turns out mypy will only complain about partial signatures when `disallow_untyped_defs` is set to `True`. I guess that makes sense, though it would be nice if there was a `disallow_partial_defs`.